File: data/lib/remote/canonical_json.py
"""Canonical JSON encoding wrapper for EFA V2 schema.

Uses the ``canonicaljson`` package (https://pypi.org/project/canonicaljson/)
to produce deterministic, minimal JSON suitable for content-addressed hashing.
"""

from __future__ import annotations

import canonicaljson

from pydantic import BaseModel


def encode_canonical_json(data: dict | list | BaseModel) -> bytes:
    """Encode *data* as canonical JSON bytes.

    Pydantic ``BaseModel`` instances are first converted to dicts with
    ``model_dump(by_alias=True)`` so that camelCase JSON keys are used.
    """
    if isinstance(data, BaseModel):
        data = data.model_dump(by_alias=True)
    return canonicaljson.encode_canonical_json(data)

File: lib/utils/canonical_json.dart
import "dart:typed_data";

import "package:canonical_json/canonical_json.dart" as cj;

/// Encodes [data] as canonical JSON bytes.
///
/// Uses the ``canonical_json`` package (https://pub.dev/packages/canonical_json)
/// to produce deterministic, minimal JSON suitable for content-addressed hashing.
Uint8List canonicalJsonEncode(Object data) => Uint8List.fromList(cj.canonicalJson.encode(data));
